retrieval/scrape.go

```diff
@@ -482,30 +498,92 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
 		}
 
 		err := sl.scraper.scrape(scrapeCtx, buf)
-		if err == nil {
-			b := buf.Bytes()
+		if err != nil && errc != nil {
+			errc <- err
+		}
 
-			if total, added, err = sl.append(b, start); err != nil {
-				log.With("err", err).Error("append failed")
+		var b []byte
+		if err == nil {
+			b = buf.Bytes()
+		}
+		// A failed scrape is the same as an empty scrape,
+		// we still call sl.append to trigger stale markers.
+		if total, added, err = sl.append(b, start); err != nil {
+			log.With("err", err).Error("append failed")
+			// The append failed, probably due to a parse error.
+			// Call sl.append again with an empty scrape to trigger stale markers.
+			if _, _, err = sl.append([]byte{}, start); err != nil {
+				log.With("err", err).Error("failure append failed")
 			}
-		} else if errc != nil {
-			errc <- err
 		}
 
 		sl.report(start, time.Since(start), total, added, err)
 		last = start
 
 		select {
 		case <-sl.ctx.Done():
+			close(sl.stopped)
 			return
+		case <-sl.scrapeCtx.Done():
+			break mainLoop
 		case <-ticker.C:
 		}
 	}
+
+	close(sl.stopped)
+
+	// Scraping has stopped. We want to write stale markers but
+	// the target may be recreated, so we wait just over 2 scrape intervals
+	// before creating them.
+	// If the context is cancelled, we presume the server is shutting down
+	// and will restart where is was. We do not attempt to write stale markers
+	// in this case.
+
+	if last.IsZero() {
+		// There never was a scrape, so there will be no stale markers.
+		return
+	}
+
+	// Wait for when the next scrape would have been, record its timestamp.
+	var staleTime time.Time
+	select {
+	case <-sl.ctx.Done():
+		return
+	case <-ticker.C:
+		staleTime = time.Now()
+	}
+
+	// Wait for when the next scrape would have been, if the target was recreated
+	// samples should have been ingested by now.
+	select {
+	case <-sl.ctx.Done():
+		return
+	case <-ticker.C:
+	}
+
+	// Wait for an extra 10% of the interval, just to be safe.
+	select {
+	case <-sl.ctx.Done():
+		return
+	case <-time.After(interval / 10):
+	}
+
+	// Call sl.append again with an empty scrape to trigger stale markers.
+	// If the target has since been recreated and scraped, the
+	// stale markers will be out of order and ignored.
```

Comment:
> Ok, so just to be clear on the rule then: staleness markers should be filtered in the adapter  for any LTS that does not support rejecting out-of-order samples?

That'd be one option for such a system where Prometheus doesn't have local storage.

> But I still don't understand why staleness wouldn't work correctly for Cortex, which does have rejection of out-of-order samples. It would at least work in most cases, right?

It'll work in some cases, not all due to additional races such a setup has.